src/ds/buffer.h

```diff
@@ -8,6 +8,51 @@
 #include <utility>
 #include <vector>
 
+struct OArray
```

Comment:
What's an OArray? Why do we need one? Why is a it struct?

Reply:
It owns an a piece of data and provides a projection onto it. By projection I mean there is a pointer that does not have to point to the start of the array and a size that does not have be the size of the whole array. 

This lets us take a buffer, process a header that exist on said data, update the projection on the array and pass it to the next layer. Since this class owns the underlying data we can pass it between threads or async calls.

Reply:
I could not find something like this in the code base. If I missed it then please tell me and I remove this class.

Reply:
Ok, thanks for adding the comments. I think it could use a couple of asserts (probably data >= d.data() also by the way).